File: tests/unit/xray/fixtures/java/pathological.java
package com.example.pathological;

import java.util.*;
import java.util.stream.*;
import java.util.function.*;

/**
 * Pathological Java: deeply nested, long lines, comment-heavy.
 * This file is intentionally complex to stress-test the tree-sitter parser.
 * It is not intended to be idiomatic or readable.
 */
public class Pathological {

    // Very long single line with deeply nested method calls and lambdas
    public static String deepNest(List<List<List<String>>> data) { return data.stream().flatMap(outer -> outer.stream().flatMap(inner -> inner.stream().filter(s -> s != null && !s.isEmpty()).map(s -> s.trim().toLowerCase().replace(" ", "_").replaceAll("[^a-z0-9_]", "")).filter(s -> s.length() > 2))).sorted(Comparator.comparingInt(String::length).thenComparing(Comparator.naturalOrder())).distinct().collect(Collectors.joining(", ", "[", "]")); }

    // Deeply nested ternary expressions
    public static int classify(int x) {
        return x < 0 ? (x < -100 ? (x < -1000 ? -4 : (x < -500 ? -3 : -2)) : -1)
                     : (x == 0 ? 0
                               : (x < 100 ? (x < 10 ? 1 : (x < 50 ? 2 : 3))
                                           : (x < 1000 ? (x < 500 ? 4 : 5) : 6)));
    }

    // Method with many parameters (stress test for parameter list parsing)
    public static Map<String, Object> buildRecord(String id, String name, String email,
            String phone, String address, String city, String state, String zip,
            String country, boolean active, int priority, double score,
            List<String> tags, Map<String, String> metadata, Date createdAt) {
        Map<String, Object> record = new LinkedHashMap<>();
        record.put("id", id);
        record.put("name", name);
        record.put("email", email);
        record.put("phone", phone);
        record.put("address", address);
        record.put("city", city);
        record.put("state", state);
        record.put("zip", zip);
        record.put("country", country);
        record.put("active", active);
        record.put("priority", priority);
        record.put("score", score);
        record.put("tags", tags);
        record.put("metadata", metadata);
        record.put("createdAt", createdAt);
        return record;
    }

    // Interface with multiple method signatures (no annotations needed)
    public interface Operation {
        double apply(double x, double y);
        default String symbol() { return "?"; }
    }

    // Implementations via lambda-style anonymous classes without annotations
    public static final Operation PLUS = (x, y) -> x + y;
    public static final Operation MINUS = (x, y) -> x - y;
    public static final Operation TIMES = (x, y) -> x * y;
    public static final Operation DIVIDE = (x, y) -> {
        if (y == 0) throw new ArithmeticException("division by zero");
        return x / y;
    };

    // Anonymous class inside lambda inside stream
    public static List<Runnable> buildTasks(List<String> names) {
        return names.stream()
                .map(name -> (Runnable) () -> {
                    // Nested comment block
                    // doing work for: name
                    String processed = name
                            .trim()
                            .toLowerCase()
                            .replace(" ", "_");
                    System.out.println("Processing: " + processed);
                })
                .collect(Collectors.toList());
    }

    // Synchronized block with nested try-catch-finally
    private static final Object LOCK = new Object();
    private static int counter = 0;

    public static int incrementSafely() {
        synchronized (LOCK) {
            try {
                // comment: increment
                counter++;
                // comment: validate
                if (counter < 0) {
                    throw new ArithmeticException("overflow");
                }
                return counter;
            } catch (ArithmeticException e) {
                counter = Integer.MAX_VALUE;
                return counter;
            } finally {
                // comment: always runs
                System.out.println("counter=" + counter);
            }
        }
    }

    // Static initializer block
    private static final Map<String, Integer> PRIORITY_MAP;
    static {
        PRIORITY_MAP = new HashMap<>();
        PRIORITY_MAP.put("critical", 1);
        PRIORITY_MAP.put("high", 2);
        PRIORITY_MAP.put("medium", 3);
        PRIORITY_MAP.put("low", 4);
        PRIORITY_MAP.put("trivial", 5);
    }

    // Nested generic types stress test
    public static <K, V extends Comparable<V>> Map<K, List<V>> groupAndSort(
            Map<K, List<V>> input) {
        Map<K, List<V>> result = new LinkedHashMap<>();
        input.forEach((key, values) -> {
            List<V> sorted = values.stream()
                    .filter(Objects::nonNull)
                    .sorted()
                    .collect(Collectors.toList());
            result.put(key, sorted);
        });
        return result;
    }

    // Chained optional operations
    public static Optional<String> processName(Optional<String> input) {
        return input
                .filter(s -> s != null)
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .map(s -> s.substring(0, 1).toUpperCase() + s.substring(1).toLowerCase())
                .filter(s -> s.matches("[A-Za-z]+"));
    }
}

File: tests/unit/xray/fixtures/kotlin/pathological.kt
package com.example.kotlin.pathological

// Deeply nested lambda chains
val deeplyNested: (List<List<List<String>>>) -> String = { data ->
    data.flatMap { outer ->
        outer.flatMap { inner ->
            inner.filter { it.isNotBlank() }
                .map { s ->
                    s.trim().lowercase()
                        .replace(Regex("[^a-z0-9]"), "_")
                        .let { slug ->
                            if (slug.length > 20) slug.substring(0, 20) else slug
                        }
                }
        }
    }.distinct().sorted().joinToString(", ")
}

// Deeply nested when expressions
fun classify(x: Int): String = when {
    x < 0 -> when {
        x < -1000 -> when {
            x < -10000 -> "very deeply negative"
            else -> "deeply negative"
        }
        x < -100 -> "moderately negative"
        else -> "slightly negative"
    }
    x == 0 -> "zero"
    else -> when {
        x > 10000 -> when {
            x > 100000 -> "astronomically positive"
            else -> "very large positive"
        }
        x > 100 -> "large positive"
        else -> "small positive"
    }
}

// Many parameters function
fun buildRecord(
    id: Long, name: String, email: String, phone: String,
    address: String, city: String, state: String, zip: String,
    country: String, active: Boolean, priority: Int, score: Double,
    tags: List<String>, metadata: Map<String, String>
): Map<String, Any?> = mapOf(
    "id" to id, "name" to name, "email" to email, "phone" to phone,
    "address" to address, "city" to city, "state" to state, "zip" to zip,
    "country" to country, "active" to active, "priority" to priority,
    "score" to score, "tags" to tags, "metadata" to metadata
)

// Long single-line chained call (stress for line length)
fun processAll(items: List<String>): List<String> = items.filter { it.isNotBlank() }.map { it.trim().lowercase() }.flatMap { it.split(",") }.map { it.trim() }.filter { it.length >= 2 }.distinct().sorted().map { it.replaceFirstChar { c -> c.uppercaseChar() } }

// Nested scope functions
fun nested(input: String?): String {
    return input?.let { raw ->
        raw.trim().let { trimmed ->
            trimmed.takeIf { it.isNotEmpty() }?.let { nonEmpty ->
                nonEmpty.also { println("Processing: $it") }
                    .run { uppercase() }
                    .also { result -> println("Result: $result") }
            } ?: "empty"
        }
    } ?: "null"
}

// Generic function with multiple bounds
fun <T> List<T>.safeGet(index: Int, default: T): T =
    if (index in indices) get(index) else default

// Recursive function
tailrec fun factorial(n: Long, acc: Long = 1L): Long =
    if (n <= 1L) acc else factorial(n - 1L, n * acc)

// String interpolation with complex expressions
fun format(items: List<Pair<String, Int>>): String =
    "Items: ${items.size}, Total: ${items.sumOf { it.second }}, " +
    "Max: ${items.maxOfOrNull { it.second } ?: 0}, " +
    "Keys: ${items.map { it.first }.sorted().joinToString("|")}"

// Anonymous object
val comparator = object : Comparator<String> {
    override fun compare(a: String, b: String): Int =
        when {
            a.length != b.length -> a.length - b.length
            else -> a.compareTo(b)
        }
}
